### ISSUE TITLE
declaration-empty-line-before

### ISSUE THREAD

**Issue by jeddy3** [stylelint/stylelint#1573]
- Name: `declaration-empty-line-before`
- Primary option: `"always"|"never"`
- Secondary options: `{ "except": ["first-nested", "after-comment", "after-declaration"] }`
- Messages: 
  - Rejected: "Unexpected empty line before declaration"
  - Expected: "Expected empty line before declaration"
- Section: "Declaration"

Given:

```js
[ "always", { "except": ["first-nested", "after-comment", "after-declaration"] }]
```

The following is valid:

```css
a {
  prop: value; /* "always" except first-nested */
  prop2: value; /* "always" except after-declaration */

  /* comment */
  prop3: value; /* "always" except after-comment */

  @extends "x";

  prop4: value; /* primary option is "always" */

   & b {
     prop: value;
   }

   prop5: value;  /* primary option is "always" */
}
```

Given:

```js
"never"
```

The following is valid:

```css
a {
  prop: value; /* primary option is "never" */
  prop2: value; /* primary option is "never" */

  /* comment */
  prop3: value; /* primary option is "never" */

  @extends "x";
  prop4: value; /* primary option is "never" */

   & b {
     prop: value;
   }
   prop5: value; /* primary option is "never" */
}
```

Ref: https://github.com/stylelint/stylelint/issues/1500#issuecomment-229300139

**Comment by evilebottnawi** [comment]
@jeddy3 you may give example?

**Comment by jeddy3** [comment]
@davidtheclark [asked](https://github.com/stylelint/stylelint/issues/1500#issuecomment-229993822): 

> If the only option here is "never", shouldn't it be declaration-no-empty-line-before?

Good question. I chose a specific `"never"` option as this rule accepts an `"except": ["first-nested"]` secondary option, which *inverts* the primary option i.e. a "first-nested" declaration *must always* have an empty line before it. e.g.

```css
a {

  display: none;
  color: red;
} 
```

However, I think you've just uncovered that there needs to be another option for when a declaration follows another node type e.g.

```css
a {
  @apply x;
  
  display: none;
  color: red;
```

Perhaps `"except": ["first"]`?

Whether it should be `declaration-empty-line-before` with a specific `"never"` option or `declaration-no-empty-line-before`, I'm not sure. We haven't had a `no` rule with an `except` secondary option before. What do you think?

**Comment by davidtheclark** [comment]
I think `declaration-no-empty-line-before` with `except` options makes more sense than a rule with only one available primary option (`"never"`).

**Comment by jeddy3** [comment]
Sounds good to me.

**Comment by evilebottnawi** [comment]
@jeddy3 if think we default ignore if first not declaration?

**Comment by jeddy3** [comment]
Updated rule details after discussion in https://github.com/stylelint/stylelint/pull/1653#issuecomment-233149070.
